### PR TITLE
feature/TECH 488 namespace logic

### PR DESCRIPTION
**Pull request by koeves**
Monorepo counterpart: https://github.com/Vandebron/Vandebron/pull/11058

- Add ci flag for health check -> if we run build health in jenkins, it reported errors that jenkins is not configured correctly
- Check for duplicate project names when we run projects lint -> will fail the build if you try to merge a duplicate project name
- Validate namespace substitution for PR target when running `mpyl project lint --extended`

Screenshot 2023-07-22 at 10 02 43

----
### 📕 [TECH-488](https://vandebron.atlassian.net/browse/TECH-488) Validate namespace replacement logic danielkoves@vandebron.nl 
Check namespace replacement

URLs should be interpolated with the name of the project

Add a command to `mpyl build health` to validate unique project names

* after loading all the projects, validating the yamls then check project names
* fail pipeline if project name is not unique
* check all project names in lower case -> give suggestions if there's a case-sensitity issue

create post about this logic + add it to mpyl documentation

🏗️ Build [7](https://jenkins.k8s-serv-backend.vdbinfra.nl/job/MPyL%20Pipeline%20-%20Test/job/PR-167/7/display/redirect) ✅ Successful, started by _Daniel Koves_  
🚀 *[cloudfront-service](https://cloudfront-service-167.test.nl/)*, *job*, *kong-sync*, *[nodeservice](https://nodeservice-167.test.nl/)*, *[sbtservice](https://sbtservice-167.test.nl/)*, *sparkJob*  

[TECH-488]: https://vandebron.atlassian.net/browse/TECH-488?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/mpyl/cli/projects.py
Comment (koeves) [reply src/mpyl/cli/projects.py:148]: Because we don't want people to merge duplicate project names (so linting should fail)

File: src/mpyl/cli/commands/projects/lint.py
Comment (koeves) [reply src/mpyl/cli/commands/projects/lint.py:82]: yeah that's absolutely right :D I had friday afternoon brain apparently

File: src/mpyl/cli/commands/projects/lint.py
Comment (koeves) [reply src/mpyl/cli/commands/projects/lint.py:29]: No because we're not constructing a list here

Comment (koeves) [reply src/mpyl/cli/commands/projects/lint.py:29]: Not sure if it's adding a lot here, we would need a nested list comprehension then an itertools chain

**Review by Jorg88 — APPROVED** [review]

File: src/mpyl/cli/commands/projects/lint.py
Comment (Jorg88) [reply src/mpyl/cli/commands/projects/lint.py]: Hm, but it finds changed projects and I feel like we've needed that before elsewhere. `.path` is just part of `Project` so that's not really relevant for this discussion I guess

Comment (koeves) [reply src/mpyl/cli/commands/projects/lint.py]: We do similar-ish things in the main mpyl flow, and when we do `mpyl build status`. There's not much to refactor though, because the changes are different per place, so you would basically only write a wrapper around find_build_set, because also its result is used differently at those places

**Review by SamTheisens — COMMENTED** [review]
@koeves, the logic is complicated quite a bit by the fact that the validation is done against a (PR) subset of projects.
Would it be possible to validate all env vars of all projects all the time instead?

**Comment by koeves** [comment]
> @koeves, the logic is complicated quite a bit by the fact that the validation is done against a (PR) subset of projects. Would it be possible to validate all env vars of all projects all the time instead?

I'd say we should only lint changed projects (like it's not really meaningful to someone that another teams project linkage is broken). Same idea as the project yaml linking before, we've had feedback that failing a build because some else committed breaking changes is not really useful in that sense. So I would keep a more complicated logic if that stops people from committing breaking changes, while keeping the error messages relevant for that person

**Comment by SamTheisens** [comment]
What is the usecase for allowing to merge broken `project.yml` to master? Is there a fundamental difference to merging code that does not compile?
Fail fast, zero tolerance is generally a saner approach IMO. You could even add a merge rule to prevent broken configurations from being merged to master. It's something that can be validated quite quickly, so I don't think it should have a negative impact on the feedback cycle.

**Comment by koeves** [comment]
> What is the usecase for allowing to merge broken `project.yml` to master? Is there a fundamental difference to merging code that does not compile? Fail fast, zero tolerance is generally a saner approach IMO. You could even create a merge rule to prevent broken configurations from being merged to master. It's something that can be validated quite quickly, so I don't think it should have a negative impact on the feedback cycle.

I meant we shouldn't validate every project on every build, because it's not relevant for the person working on some projects. We still validate the changed projects and fail fast if they are invalid.

**Comment by SamTheisens** [comment]
Why not validate everything all the time? That way you can't have broken `project.yml` in master by definition. And you don't need to even think about this complicating scenario.

**Comment by Jorg88** [comment]
I agree with you Sam, but I lost the discussion as you can see 😉 important detail though: we came to a comprimise that for now we only lint changed projects since breaking changes can still be merged (and build via MPL), and we would look into maybe a github action to block merging breaking changes after the migration is complete. After that we can lint/validate everything all the time

**Comment by koeves** [comment]
> Why not validate everything all the time? That way you can't have broken `project.yml` in master by definition. And you don't need to even think about this complicating scenario.

We can, but until everyone is using MPyL people will merge broken yamls.. All it takes is one project for your build to fail, and then what do you do? Do you correct another teams project in your PR? You're basically blocked until they fix it, or you spend time fixing others' projects.

**Comment by koeves** [comment]
I added the option that we lint all projects for everything. It looks like the namespace linkup fails for the projects that use the project overrides.. This is not ideal because we didn't add the project overrides to MPyL. Also came up in a meeting that teams will get rid of this functionality sometime. So for now we can stick to linting changed projects
Screenshot 2023-07-26 at 20 19 13

**Comment by SamTheisens** [comment]
@koeves I see 🤔. In that case maybe it would be better to exclude these projects altogether? Make them effectively invisible to MpyL I mean.
We don't want them to be included in builds either, as long as we don't support overrides.
As a temporary hack we could filter out a list of projects (perhaps via csv env var) here https://github.com/Vandebron/mpyl/blob/211775ceb648d7b266b19bf627cbb9e67705d8be/src/mpyl/utilities/repo/__init__.py#L223-L232

I feel that would be a simpler temporary hack than looking at whatever project has changed.

For the override stuff we could perhaps take inspiration from (or just use) the application sets in argocd https://argo-cd.readthedocs.io/en/stable/operator-manual/applicationset/.

**Pull request merged**
